x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java

```diff
@@ -608,35 +630,37 @@ protected Clock getClock() {
                 new ActionHandler<>(SetUpgradeModeAction.INSTANCE, TransportSetUpgradeModeAction.class)
         );
     }
+
     @Override
     public List<ExecutorBuilder<?>> getExecutorBuilders(Settings settings) {
         if (false == enabled || transportClientMode) {
             return emptyList();
         }
-        int maxNumberOfJobs = AutodetectProcessManager.MAX_OPEN_JOBS_PER_NODE.get(settings);
-        // 4 threads per job: for cpp logging, result processing, state processing and
-        // AutodetectProcessManager worker thread:
-        FixedExecutorBuilder autoDetect = new FixedExecutorBuilder(settings, AUTODETECT_THREAD_POOL_NAME,
-                maxNumberOfJobs * 4, maxNumberOfJobs * 4, "xpack.ml.autodetect_thread_pool");
-
-        // 4 threads per job: processing logging, result and state of the renormalization process.
-        // Renormalization does't run for the entire lifetime of a job, so additionally autodetect process
-        // based operation (open, close, flush, post data), datafeed based operations (start and stop)
-        // and deleting expired data use this threadpool too and queue up if all threads are busy.
-        FixedExecutorBuilder renormalizer = new FixedExecutorBuilder(settings, UTILITY_THREAD_POOL_NAME,
-                maxNumberOfJobs * 4, 500, "xpack.ml.utility_thread_pool");
-
-        // TODO: if datafeed and non datafeed jobs are considered more equal and the datafeed and
-        // autodetect process are created at the same time then these two different TPs can merge.
-        FixedExecutorBuilder datafeed = new FixedExecutorBuilder(settings, DATAFEED_THREAD_POOL_NAME,
-                maxNumberOfJobs, 200, "xpack.ml.datafeed_thread_pool");
-        return Arrays.asList(autoDetect, renormalizer, datafeed);
+
+        // These thread pools scale such that they can accommodate the maximum number of worker processes per node
+        // that is permitted to be configured.  It is up to other code to enforce the configured maximum number of
+        // worker processes per node.
+
+        // 4 threads per worker process: for input, c++ logger output, result processing and state processing.
+        ScalingExecutorBuilder workerComms = new ScalingExecutorBuilder(WORKER_COMMS_THREAD_POOL_NAME,
+            4, MAX_MAX_OPEN_WORKERS_PER_NODE * 4, TimeValue.timeValueMinutes(1), "xpack.ml.worker_comms_thread_pool");
+
+        // This pool is used by renormalization, plus some other parts of ML that
+        // need to kick off non-trivial activities that mustn't block other threads.
+        ScalingExecutorBuilder utility = new ScalingExecutorBuilder(UTILITY_THREAD_POOL_NAME,
+            1, MAX_MAX_OPEN_WORKERS_PER_NODE * 4, TimeValue.timeValueMinutes(10), "xpack.ml.utility_thread_pool");
```

Comment:
It's not just renormalization that uses this thread pool - it's all these places:

```
src/main/java/org/elasticsearch/xpack/ml/action/TransportKillProcessAction.java:            KillProcessAction.Response::new, MachineLearning.UTILITY_THREAD_POOL_NAME, processManager);
src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java:            threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(new AbstractRunnable() {
src/main/java/org/elasticsearch/xpack/ml/action/TransportGetOverallBucketsAction.java:                    threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
src/main/java/org/elasticsearch/xpack/ml/action/TransportFindFileStructureAction.java:        threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
src/main/java/org/elasticsearch/xpack/ml/action/TransportFinalizeJobExecutionAction.java:                MachineLearning.UTILITY_THREAD_POOL_NAME), true);
src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java:            StopDatafeedAction.Response::new, StopDatafeedAction.Response::new, MachineLearning.UTILITY_THREAD_POOL_NAME);
src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java:                    threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(new AbstractRunnable() {
src/main/java/org/elasticsearch/xpack/ml/action/TransportOpenJobAction.java:            super(MlTasks.JOB_TASK_NAME, MachineLearning.UTILITY_THREAD_POOL_NAME);
src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteExpiredDataAction.java:        threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> deleteExpiredData(listener));
src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteExpiredDataAction.java:            remover.remove(new ThreadedActionListener<>(logger, threadPool, MachineLearning.UTILITY_THREAD_POOL_NAME, nextListener,
src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedAction.java:            super(MlTasks.DATAFEED_TASK_NAME, MachineLearning.UTILITY_THREAD_POOL_NAME);
src/main/java/org/elasticsearch/xpack/ml/action/TransportIsolateDatafeedAction.java:            IsolateDatafeedAction.Response::new, IsolateDatafeedAction.Response::new, MachineLearning.UTILITY_THREAD_POOL_NAME);
src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java:            super(MlTasks.DATA_FRAME_ANALYTICS_TASK_NAME, MachineLearning.UTILITY_THREAD_POOL_NAME);
src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredForecastsRemover.java:                MachineLearning.UTILITY_THREAD_POOL_NAME, forecastStatsHandler, false));
src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredModelSnapshotsRemover.java:                MachineLearning.UTILITY_THREAD_POOL_NAME, expiredSnapshotsListener(job.getId(), listener), false));
src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java:                MachineLearning.UTILITY_THREAD_POOL_NAME), true);
src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java:                    threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java:                    threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java:                            threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(new AbstractRunnable() {
src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java:        ExecutorService renormalizerExecutorService = threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME);
src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java:        return MachineLearning.UTILITY_THREAD_POOL_NAME;
```

We don't want the threads in this thread pool to be stopped too aggressively, because that would mean that every one of those places has to start a new OS thread when run.

The idea behind the 10 minutes was to very gradually adapt to the average concurrency of those methods.

Maybe an alternative would be to stop unused threads more aggressively, but keep a core of, say, 25, threads all the time.  What do you think?

Reply:
If ml isn't being used it seems a shame to grab 25 threads and defeat the purpose of the scaling executor. I hate to speculate about the performance characteristics of the various options but 10 minutes seems a reasonable choice given the above